File: python/n501_BST.py
# https://leetcode-cn.com/problems/find-mode-in-binary-search-tree/
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def findMode(self, root: TreeNode):
        lis = []
        if root is None: return lis
        self.TransTree(root, lis)
        res = []
        max_num = 0
        cur_num = 1
        l = len(lis)
        for i in range(1, l+1):
            if i == l or lis[i] != lis[i-1]:
                if cur_num > max_num:
                    res.clear()
                    res.append(lis[i-1])
                    max_num = cur_num
                elif cur_num == max_num:
                    res.append(lis[i-1])
                cur_num = 1
            else:
                cur_num += 1
        return res                
    
    def TransTree(self, root: TreeNode, lis):
        if root is None: return
        self.TransTree(root.left, lis)
        lis.append(root.val)
        self.TransTree(root.right, lis)
        


if __name__ == "__main__":
    pass
        

File: cpp/n501_BST.cpp
// https://leetcode-cn.com/problems/find-mode-in-binary-search-tree/
// Definition for a binary tree node.
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<int> findMode(TreeNode* root) {
        vector<int> lis;
        if (root == NULL) return lis;
        TransTree(root, lis);
        vector<int> res;
        int max_count = 0;
        int cur_count = 1;
        for (size_t i = 1; i <= lis.size(); ++i){
            if (i == lis.size() || lis[i] != lis[i-1]){
                if (cur_count > max_count){
                    res.clear();
                    res.push_back(lis[i-1]);
                    max_count = cur_count;
                }
                else if (cur_count == max_count) res.push_back(lis[i-1]);
                cur_count = 1;
            }
            else cur_count += 1;
        }
        return res;
    }
    
    void TransTree(TreeNode* root, vector<int>& lis){
        if (root == NULL) return;
        TransTree(root -> left, lis);
        lis.push_back(root -> val);
        TransTree(root -> right, lis);
    }
};

int main(){
    return 0;
}